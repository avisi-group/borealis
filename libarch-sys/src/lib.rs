#![allow(non_upper_case_globals)]
#![allow(non_camel_case_types)]
#![allow(non_snake_case)]

include!(concat!(env!("OUT_DIR"), "/bindings.rs"));

use {
    crate::root::{
        captive::arch::arm64::{arm64_decode_decode, arm64_disasm_disassemble},
        new_decoder, new_disassembler,
    },
    std::{
        ffi::{c_void, CStr},
        fmt::Write,
    },
};

pub fn disassemble(data: &[u8]) -> String {
    let mut decode_data = [0u8; 128];
    let mut disasm_data = [0u8; 128];

    let decoder = unsafe { new_decoder(&mut decode_data as *mut u8 as *mut c_void) };
    let disassembler = unsafe { new_disassembler(&mut disasm_data as *mut u8 as *mut c_void) };

    let mut output = String::new();

    for chunk in data.chunks_exact(4) {
        unsafe { arm64_decode_decode(decoder as *mut c_void, 0, 0, &chunk[0]) };

        let str_ptr = unsafe {
            arm64_disasm_disassemble(disassembler as *mut c_void, 0, decoder as *const u8)
        };

        writeln!(
            output,
            "{:08X}: {}",
            unsafe { std::mem::transmute::<_, &u32>(&chunk[0]) },
            unsafe { CStr::from_ptr(str_ptr) }.to_str().unwrap()
        )
        .unwrap();
    }

    output
}

#[cfg(test)]
mod tests {
    #[test]
    fn decode() {
        // Aarch64 assembly for
        //
        // ```
        // #include <unistd.h>
        //
        // int main()
        // {
        //     const char msg[] = "Hello, ARM!\n";
        //     __uint32_t a = 123 + 37;
        //     __uint32_t b = a * 7;
        //     __uint32_t c = a / b;
        //     write(0, msg, sizeof(msg));
        //     _exit(0);
        // }
        // ```
        let data = [
            0xFD, 0x7B, 0xBC, 0xA9, 0xFD, 0x03, 0x00, 0x91, 0x00, 0x00, 0x00, 0x90, 0x00, 0x00,
            0x40, 0xF9, 0x01, 0x00, 0x40, 0xF9, 0xE1, 0x1F, 0x00, 0xF9, 0x01, 0x00, 0x80, 0xD2,
            0x00, 0x00, 0x00, 0x90, 0x01, 0x00, 0x00, 0x91, 0xE0, 0xA3, 0x00, 0x91, 0x22, 0x00,
            0x40, 0xF9, 0x02, 0x00, 0x00, 0xF9, 0x21, 0x50, 0x40, 0xF8, 0x01, 0x50, 0x00, 0xF8,
            0x00, 0x14, 0x80, 0x52, 0xE0, 0x1F, 0x00, 0xB9, 0xE1, 0x1F, 0x40, 0xB9, 0xE0, 0x03,
            0x01, 0x2A, 0x00, 0x70, 0x1D, 0x53, 0x00, 0x00, 0x01, 0x4B, 0xE0, 0x23, 0x00, 0xB9,
            0xE1, 0x1F, 0x40, 0xB9, 0xE0, 0x23, 0x40, 0xB9, 0x20, 0x08, 0xC0, 0x1A, 0xE0, 0x27,
            0x00, 0xB9, 0xE0, 0xA3, 0x00, 0x91, 0xA2, 0x01, 0x80, 0xD2, 0xE1, 0x03, 0x00, 0xAA,
            0x00, 0x00, 0x80, 0x52, 0x00, 0x00, 0x00, 0x94, 0x00, 0x00, 0x80, 0x52, 0x00, 0x00,
            0x00, 0x94,
        ];

        insta::assert_snapshot!(crate::disassemble(&data));
    }
}
