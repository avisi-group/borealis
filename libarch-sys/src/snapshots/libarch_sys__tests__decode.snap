---
source: libarch-sys/src/lib.rs
expression: "crate::disassemble(&data)"
---
A9BC7BFD: memory_pair_general_preidx_memory_pair_general_postidx__decode Rt: 0x1, 5, 6, Rt2: 0x1e, imm7: 0x1e, L: 0x1, padding1446: 0x7, V: 0x1, padding1447: 0x7, opc: 0x1, 
910003FD: integer_arithmetic_addsub_immediate_decode Rd: 0x1, 2, 1, Rn_part1: 0x0, imm12: 0x0, shift: 0x3, padding130: 0x1f, S: 0x1, op: 0x0, sf: 0x1, 
90000000: integer_arithmetic_address_pcrel_decode 1, Rd_part1: 0x0, 2, 0, immhi_part1: 0x0, padding1275: 0x0, immlo: 0x0, op: 0x0, 
F9400000: memory_single_general_immediate_unsigned_memory_single_general_immediate_signed_postidx__decode 1, Rt_part1: 0x1, 7, 5, imm12: 0x0, opc: 0x0, padding1758: 0x0, V: 0x0, padding1759: 0x0, size: 0x0, 
F9400001: memory_single_general_immediate_unsigned_memory_single_general_immediate_signed_postidx__decode 1, Rt_part1: 0x1, 7, 5, imm12: 0x0, opc: 0x0, padding1758: 0x0, V: 0x0, padding1759: 0x0, size: 0x1, 
F9001FE1: memory_single_general_immediate_unsigned_memory_single_general_immediate_signed_postidx__decode 1, Rt_part1: 0x1, 7, 4, imm12: 0x7, opc: 0x3, padding865: 0x3, V: 0x1, padding866: 0x0, size: 0x1, 
D2800001: integer_insext_insert_movewide_decode Rd: 0x1, 10, 5, imm16_part1: 0x0, hw: 0x0, padding1599: 0x0, opc: 0x0, sf: 0x1, 
90000000: integer_arithmetic_address_pcrel_decode 1, Rd_part1: 0x0, 2, 0, immhi_part1: 0x0, padding1275: 0x0, immlo: 0x0, op: 0x0, 
91000001: integer_arithmetic_addsub_immediate_decode Rd: 0x1, 2, 1, Rn_part1: 0x0, imm12: 0x0, shift: 0x0, padding130: 0x0, S: 0x0, op: 0x0, sf: 0x1, 
9100A3E0: integer_arithmetic_addsub_immediate_decode Rd: 0x1, 2, 1, Rn_part1: 0x0, imm12: 0x28, shift: 0x3, padding130: 0x1c, S: 0x0, op: 0x0, sf: 0x0, 
F9400022: memory_single_general_immediate_unsigned_memory_single_general_immediate_signed_postidx__decode 1, Rt_part1: 0x1, 7, 5, imm12: 0x0, opc: 0x0, padding1758: 0x0, V: 0x1, padding1759: 0x0, size: 0x2, 
F9000002: memory_single_general_immediate_unsigned_memory_single_general_immediate_signed_postidx__decode 1, Rt_part1: 0x1, 7, 4, imm12: 0x0, opc: 0x0, padding865: 0x0, V: 0x0, padding866: 0x0, size: 0x2, 
F8405021: memory_single_general_immediate_signed_offset_normal_memory_single_general_immediate_signed_offset_normal__decode 1, Rt_part1: 0x1, 7, 1, 0, padding534_part1: 0x0, imm9: 0x5, 0, 0, opc: 0x0, padding536: 0x0, V: 0x1, padding537: 0x0, size: 0x1, 
F8005001: memory_single_general_immediate_signed_offset_normal_memory_single_general_immediate_signed_offset_normal__decode 1, Rt_part1: 0x1, 7, 0, 0, padding287_part1: 0x0, imm9: 0x5, 0, 0, opc: 0x0, padding289: 0x0, V: 0x0, padding290: 0x0, size: 0x1, 
52801400: integer_insext_insert_movewide_decode Rd: 0x0, 10, 5, imm16_part1: 0x2, hw: 0x2, padding1599: 0x0, opc: 0x0, sf: 0x0, 
B9001FE0: memory_single_general_immediate_unsigned_memory_single_general_immediate_signed_postidx__decode 1, Rt_part1: 0x0, 7, 4, imm12: 0x7, opc: 0x3, padding865: 0x3, V: 0x1, padding866: 0x0, size: 0x0, 
B9401FE1: memory_single_general_immediate_unsigned_memory_single_general_immediate_signed_postidx__decode 1, Rt_part1: 0x0, 7, 5, imm12: 0x7, opc: 0x3, padding1758: 0x3, V: 0x1, padding1759: 0x0, size: 0x1, 
2A0103E0: integer_logical_shiftedreg_decode Rd: 0x0, 5, 2, Rn_part1: 0x0, 0, imm6_part1: 0x1, Rm: 0x0, N: 0x0, shift: 0x3, padding2891: 0x1c, opc: 0x0, sf: 0x0, 
531D7000: integer_bitfield_decode Rd: 0x0, 10, 6, Rn_part1: 0x0, imms: 0x1d, immr: 0x1c, N: 0x0, padding2592: 0x0, opc: 0x0, sf: 0x0, 
4B010000: integer_arithmetic_addsub_shiftedreg_decode Rd: 0x0, 9, 3, Rn_part1: 0x0, 0, imm6_part1: 0x1, Rm: 0x0, padding2: 0x0, shift: 0x0, padding3: 0x0, S: 0x0, op: 0x0, sf: 0x0, 
B90023E0: memory_single_general_immediate_unsigned_memory_single_general_immediate_signed_postidx__decode 1, Rt_part1: 0x0, 7, 4, imm12: 0x8, opc: 0x3, padding865: 0x3, V: 0x1, padding866: 0x0, size: 0x0, 
B9401FE1: memory_single_general_immediate_unsigned_memory_single_general_immediate_signed_postidx__decode 1, Rt_part1: 0x0, 7, 5, imm12: 0x7, opc: 0x3, padding1758: 0x3, V: 0x1, padding1759: 0x0, size: 0x1, 
B94023E0: memory_single_general_immediate_unsigned_memory_single_general_immediate_signed_postidx__decode 1, Rt_part1: 0x0, 7, 5, imm12: 0x8, opc: 0x3, padding1758: 0x3, V: 0x1, padding1759: 0x0, size: 0x0, 
1AC00820: integer_arithmetic_div_decode Rd: 0x0, 3, 11, 0, opcode2_5_1_: 0x0, 1, 0, padding3072_part1: 0x4, S: 0x0, op: 0x0, sf: 0x0, 
B90027E0: memory_single_general_immediate_unsigned_memory_single_general_immediate_signed_postidx__decode 1, Rt_part1: 0x0, 7, 4, imm12: 0x9, opc: 0x3, padding865: 0x3, V: 0x1, padding866: 0x0, size: 0x0, 
9100A3E0: integer_arithmetic_addsub_immediate_decode Rd: 0x1, 2, 1, Rn_part1: 0x0, imm12: 0x28, shift: 0x3, padding130: 0x1c, S: 0x0, op: 0x0, sf: 0x0, 
D28001A2: integer_insext_insert_movewide_decode Rd: 0x1, 10, 5, imm16_part1: 0x0, hw: 0x0, padding1599: 0x34, opc: 0x1, sf: 0x0, 
AA0003E1: integer_logical_shiftedreg_decode Rd: 0x1, 5, 2, Rn_part1: 0x0, 0, imm6_part1: 0x0, Rm: 0x0, N: 0x0, shift: 0x3, padding2891: 0x1c, opc: 0x0, sf: 0x1, 
52800000: integer_insext_insert_movewide_decode Rd: 0x0, 10, 5, imm16_part1: 0x0, hw: 0x0, padding1599: 0x0, opc: 0x0, sf: 0x0, 
94000000: branch_unconditional_immediate_decode 37, imm26_part1: 0x0, padding2006: 0x0, op: 0x0, 
52800000: integer_insext_insert_movewide_decode Rd: 0x0, 10, 5, imm16_part1: 0x0, hw: 0x0, padding1599: 0x0, opc: 0x0, sf: 0x0, 
94000000: branch_unconditional_immediate_decode 37, imm26_part1: 0x0, padding2006: 0x0, op: 0x0, 

